### PR TITLE
EREP-010: paymaster should have deposit to cover all userops

### PR DESCRIPTION
**Pull request by drortirosh**
submitting a new UserOp into the mempool should revert if its deposit is not high enough.

### CODE REVIEW

File: packages/bundler/src/modules/DepositManager.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    // on-chain AA31 checks the deposit for the current userop.
+    // but submitting all these UserOps it will eventually abort on this error,
+    // so it's fine to return the same code.
+    requireCond(deposit.gte(0), 'paymaster deposit too low for all mempool UserOps', ValidationErrors.BannedOrThrottledPaymaster)
+  }
+
+  /**
+   * clear deposits after some known change on-chain
+   */
+  clearCache (): void {
```
Comment (forshtat) [inline packages/bundler/src/modules/DepositManager.ts:40]: I think it is a bad idea to have the cache stored in `DepositManager` but trigger the clearing of the cache from the `EventsManager`. I suggest you keep the entire logic of the cache inside this class, i.e. let the `getCachedDeposit()` function do the `clearCache` once every X blocks or whatever.

File: packages/bundler/src/modules/MempoolManager.ts [hunk 2]
```diff
@@ -92,15 +92,15 @@ export class MempoolManager {
       this.mempool[index] = entry
     } else {
       debug('add userOp', userOp.sender, userOp.nonce)
-      this.incrementEntryCount(userOp.sender)
+      this.checkReputation(senderInfo, paymasterInfo, factoryInfo, aggregatorInfo)
+      this.checkMultipleRolesViolation(userOp)
       if (userOp.paymaster != null) {
         this.incrementEntryCount(userOp.paymaster)
       }
+      this.incrementEntryCount(userOp.sender)
```
Comment (forshtat) [inline packages/bundler/src/modules/MempoolManager.ts:100]: Why move the `incrementEntryCount(sender)` to be below `incrementEntryCount(paymaster)`? Seems like an unnecessary change to me.

File: packages/bundler/src/modules/MempoolManager.ts [hunk 3]
```diff
@@ -290,4 +290,8 @@ export class MempoolManager {
 
     return res.filter(it => it != null).map(it => (it as string).toLowerCase())
   }
+
+  getMempool (): MempoolEntry[] {
```
Comment (forshtat) [inline packages/bundler/src/modules/MempoolManager.ts:294]: I am concerned that the `removeUserOp` function is only triggered when bundler attempts to include a transaction and it fails. Before this PR the `mempool` was basically private so this did not matter. But now, you rely on the mempool to only have valid Paymaster transactions to decide if you are adding a new UserOp or not. 
Practically, your mempool may be choke-full of invalid Paymaster transactions that it never gets to do a second validation for, and therefore refuses to add valid UserOps for this Paymaster instead.

File: packages/bundler/test/ValidateManager.test.ts [hunk 4]
```diff
@@ -331,7 +331,7 @@ describe('#ValidationManager', () => {
     await testExistingUserOp('balance-self', undefined)
   })
 
-  it('should fail with unstaked paymaster returning context', async () => {
+  it('should accept unstaked paymaster returning context', async () => {
```
Comment (forshtat) [inline packages/bundler/test/ValidateManager.test.ts:334]: Is this change supposed to be included in this PR?

File: packages/utils/src/Utils.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ export enum ValidationErrors {
   SimulatePaymasterValidation = -32501,
   OpcodeValidation = -32502,
   NotInTimeRange = -32503,
-  Reputation = -32504,
+  BannedOrThrottledPaymaster = -32504,
```
Comment (forshtat) [inline packages/utils/src/Utils.ts:36]: I would rather split these into two separate error codes.

File: packages/utils/src/Utils.ts [hunk 6]
```diff
@@ -189,3 +189,19 @@ export async function runContractScript<T extends ContractFactory> (provider: Pr
...
+export function sum (...args: BigNumberish[]): BigNumber {
+  return args.reduce((acc: BigNumber, cur) => acc.add(cur), BigNumber.from(0))
+}
+
+/**
+ * calculate the maximum verification cost of a UserOperation.
+ * the cost is the sum of the verification gas limits, multiplied by the maxFeePerGas.
+ * @param userOp
+ */
+export function getUserOpMaxCost (userOp: UserOperation): number {
```
Comment (forshtat) [inline packages/utils/src/Utils.ts:205]: This looks like Go code. Shouldn't we make `UserOperation` a class?

**Review by forshtat — CHANGES_REQUESTED** [review]

File: packages/bundler/src/modules/DepositManager.ts (hunk 1 above)
Comment (drortirosh) [reply packages/bundler/src/modules/DepositManager.ts:40]: the DepositManager manages the deposits. the simplest impl is to read from the blockchain everytime.
Instead, I wanted some optimization, and cache it until there is a network change - which is always signaled by a message.

File: packages/bundler/src/modules/MempoolManager.ts (hunk 2 above)
Comment (drortirosh) [reply packages/bundler/src/modules/MempoolManager.ts:100]: don't increment before you solved all cases where it can crash (otherwise, you need to catch to undo increments)

File: packages/bundler/src/modules/MempoolManager.ts (hunk 3 above)
Comment (drortirosh) [reply packages/bundler/src/modules/MempoolManager.ts:294]: its a separate bug: removeUserOp should be triggered only by events, and not after submitting a bundle (should be opened as a separate bug, unrelated to this erep)

File: packages/bundler/test/ValidateManager.test.ts (hunk 4 above)
Comment (drortirosh) [reply packages/bundler/test/ValidateManager.test.ts:334]: unrelated. will separate.

File: packages/utils/src/Utils.ts (hunk 6 above)
Comment (drortirosh) [reply packages/utils/src/Utils.ts:205]: yes, should pack all userop related utils (resemble UserOperationLib in solidity - but we don't have "using" in TS..)

**Review by forshtat — APPROVED** [review]
